### DISCUSSION TITLE
[Nuxt & Nitro] Nitro Plugin Context & Nuxt Plugin Context

### DISCUSSION THREAD

**Discussion by Denoder** [nuxt/nuxt#24019] [General]
I'm kind of confused as to the working of nitro with Nuxt. With Nuxt 2, you had a server layer and a client layer. You separate that disparity with something like `proces.client` or `process.server`.

With Nuxt 3 you have a Nuxt plugin which also utilizes the same principles as Nuxt 2, but with an additional layer or what I'd like to believe is an additional layer, which is nitro's server, which is inaccessible in Nuxt's context. It feels like with:

Nuxt 2 you went like this:
`Sever -> Client`

Nuxt 3 feels like:
`Nitro Server -> Server (?) -> Client`

With that setup in mind, it feels like an unnecessary step is being added to Nuxt, because you have nitro plugins which cater to the nitro setup and has limited to no access to Nuxt's layer. Then on top of that you have Nuxt plugins which does have some semblance of server access, but not towards nitro because if you're using some like a server middleware you can't just use `process.server` in the plugin because it won't be available in nitro's lifecycle since it runs before Nuxt initializes nor does any of Nuxt's features work in the server middleware. 

Likewise things you do in nitro doesn't translate over to Nuxt's side all that well, also you kind of have to create 2 variants of the same plugin to handle both nitro and Nuxt where both of them do the same thing. On top of this there's 2 layers of middleware here; Nuxt middleware, and Nitro's server middleware, both of which doe what middlewares do, but the Nuxt middleware also accepts `process.sever` but also doesn't have access to nitro?

It's a bit confusing to have this kind of setup.

**Comment by zestlee1106** [comment]
I agree with you.

I am struggling because useState and pinia are not available while configuring server middleware.

Importantly, the Nuxt3 document does not accurately point out that Nitro and Nuxt servers should be thought of separately.

**Reply by keyss2014** [reply]
i found NuxtApp is isolated from  NitroApp , I setup a global variable in nitro server, but can't get the value from Nuxt APP server side!

**Comment by TheAlexLichter** [comment]
Here is [an overview of what Nitro does vs. what Nuxt does](https://www.youtube.com/watch?v=DkvgJa-X31k).

The separation between "API routes" that don't need Nuxt to be ran at all, versus Nitro "booting up" the Nuxt application, which then might server-render or not (depending on settings) is important to separate concerns, enable features like cached functions easier etc.

**Comment by jmarshall9120** [comment]
@TheAlexLichter - I'm having an issue here where I'm developing a module and need to inject a static variable into the event context for a handler that my module adds (addServerHandler or addDevHandler).  There seems to be some life cycle hang up here.  Injecting variables to run time context like:

```
    const config = useRuntimeConfig();
    config.public.__MYPUBLICVAL__ = 'M://source/myProject/tests/mocks';
    config.__MYPRIVATEVAL__ = 'M://source/myProject/tests/mocks';
```
or
```
nuxt.options.nitro.runtimeConfig.__MYPUBLICVAL__ = 'M://source/myProject/tests/mocks';
```

Don't appear at runtime in the added handler's event context.  I'm confused about how a module should inject necessary context variables?

**Reply by TheAlexLichter** [reply]
A repro would be appreciated!

**Reply by jmarshall9120** [reply]
here's a repo:
https://wmaioaxagygithub-q4wp--3000--96435430.local.webcontainer.io

Just look at the server output when you start the project.  Expected output would be to get the same path at handler registration as at runtime.

**Reply by jmarshall9120** [reply]
@TheAlexLichter  - I just realized that link doesn't share the playground as I would expect.  I've downloaded it and shared the repo here:

https://github.com/jmarshall9120/nuxt_module_context_demo
